src/libraries/Microsoft.Extensions.Logging.Abstractions/tests/Microsoft.Extensions.Logging.Generators.Tests/TestClasses/ConstaintsTestExtensions.cs

```diff
@@ -0,0 +1,98 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+namespace Microsoft.Extensions.Logging.Generators.Tests.TestClasses
+{
+    using ConstraintInAnotherNamespace;
+    namespace UsesConstraintInAnotherNamespace
+    {
+        public partial class MessagePrinter<T>
+            where T : Message
+        {
+            public void Print(ILogger logger, T message)
+            {
+                Log.Message(logger, message.Text);
+            }
+
+            internal static partial class Log
+            {
+                [LoggerMessage(EventId = 1, Level = LogLevel.Information, Message = "The message is {Text}.")]
+                internal static partial void Message(ILogger logger, string? text);
+            }
+        }
+    }
+
+    internal static partial class ConstraintsTestExtensions<T>
+        where T : class
+    {
+        [LoggerMessage(EventId = 0, Level = LogLevel.Debug, Message = "M0{p0}")]
+        public static partial void M0(ILogger logger, int p0);
+
+        public static void Foo(T dummy)
+        {
+        }
+    }
+
+    internal static partial class ConstraintsTestExtensions1<T>
+        where T : struct
+    {
+        [LoggerMessage(EventId = 0, Level = LogLevel.Debug, Message = "M0{p0}")]
+        public static partial void M0(ILogger logger, int p0);
+
+        public static void Foo(T dummy)
+        {
+        }
+    }
+
+    internal static partial class ConstraintsTestExtensions2<T>
+        where T : unmanaged
+    {
+        [LoggerMessage(EventId = 0, Level = LogLevel.Debug, Message = "M0{p0}")]
+        public static partial void M0(ILogger logger, int p0);
+
+        public static void Foo(T dummy)
+        {
+        }
+    }
+
+    internal static partial class ConstraintsTestExtensions3<T>
+        where T : new()
+    {
+        [LoggerMessage(EventId = 0, Level = LogLevel.Debug, Message = "M0{p0}")]
+        public static partial void M0(ILogger logger, int p0);
+
+        public static void Foo(T dummy)
+        {
+        }
+    }
+
+    internal static partial class ConstraintsTestExtensions4<T>
+        where T : System.Attribute
+    {
+        [LoggerMessage(EventId = 0, Level = LogLevel.Debug, Message = "M0{p0}")]
+        public static partial void M0(ILogger logger, int p0);
+
+        public static void Foo(T dummy)
+        {
+        }
+    }
+
+    internal static partial class ConstraintsTestExtensions5<T>
+        where T : notnull
+    {
+        [LoggerMessage(EventId = 0, Level = LogLevel.Debug, Message = "M0{p0}")]
```

Comment:
Is it possible to use a variable of type `T` as a parameter in the partial method?  If so that might be a good test to add.  If we don't allow it then it's OK to omit.

Reply:
Added test.